sycl/doc/design/SYCL-1Pass.md

```diff
@@ -0,0 +1,57 @@
+# SYCL 1-Pass Compilation Model
+
+Currently the DPC++ compiler utilizes a scheme which needs multiple invocations of the compiler.  Specifically, it invokes the cc1 compiler once for each device compilation and one more time for the host compilation.  This document describes design changes necessary (in the various components of the DPC++ Toolchain) to switch to a single invocation of the cc1 compiler for each Translation Unit (TU).
+
+## Compiler Front End (FE) Changes
+
+This design reuqires some changes from the FE.  In particular, it needs to emit code via the normal/standard C++ rules, and in addition, needs to emit functions that are required for Kernel code, which may not have been emitted via the normal C++ rules.
+
+### Generation and dispatch to the OpenCL Kernel
+
+The FE will generate the OpenCL Kernel which creates the Functor object that it will call into.  FE then replaces the [[sycl_kernel]] function with one that dispatches to the OpenCL Kernel.  New APIs from the SYCL Runtime Library (RTL) are needed to accomplish this task.
+
+
+### Elimination of generation of the integration header file
+
+Since there is no longer separate device and host compilation invocations, communication between them via the integration header file is no longer necessary.  Instead the FE can pass the information previously generated in the header file directly to the RTL via new APIs.
+
+### Elimination of generation of the integration footer file
+Similarly, an integration footer file too is no longer necessary, but specialization constants need to still be handled.
+
+### Intrinsics and builtins
+There will need to be some kind of collection of all intrinsics that we support over all devices into one table.  If intrinsics are spelled the same but have differing signatures, they will need to be renamed to be unique across the affected devices.
+
+### Function attributes and instruction extensions
+TODO
+
+### Deferred device diagnostics
+FE will add an end-of-TU checker to emit deferred device diagnostics.
```

Comment:
A vast majority of our diagnostics will have to be converted to be exclusively deferred-diagnostics when possible.  Additionally, any post-Sema diagnostics (which, are likely candidates for removal...) DO need to be done.

Reply:
`Additionally, any post-Sema diagnostics (which, are likely candidates for removal...) DO need to be done.`

Could you clarify what you mean here, @erichkeane?

Reply:
so we have what I consider to be 3 classes of diagnostics these days:

1- Currently Deferred Diagnostics
2- Diagnostics done 'after' Sema (during SemaSYCL's kernel generation) that SHOULD be Deferred Diagnostics.
3- Diagnostics done 'after' Sema that aren't possible during Deferred Diagnostics.

We want to keep 1, and convert 2 as a part of this.  Finally, the ones in 3 need to be re-evaluated as to whether we should be keeping them at all (or replacing the language rule with something else).  If it is an absolute NEED and we can convince community of it, we need to do them as well.